Выполнено практическое задание 9-го спринта.

MaximusDenisov-YP opened the pull request:
- Добавлено HTTP взаимодействие с приложением.
- Произведён маппинг методов.
- Добавлены Handlers and Adapters для сущностей.
- Добавлены утилитарные классы GsonUtil and JsonTypes.
- Добавлены тесты по написанным Handlers.

Commit:
Выполнено практическое задание 9-го спринта.

- Добавлено HTTP взаимодействие с приложением.
- Произведён маппинг методов.
- Добавлены Handlers and Adapters для сущностей.
- Добавлены утилитарные классы GsonUtil and JsonTypes.
- Добавлены тесты по написанным Handlers.

Commit:
Исправил замечание по CheckStyle JLS.

avfyodorov commented on src/main/ru/kanban/http/handler/EpicsHandler.java:
```diff
@@ -0,0 +1,105 @@
+
+public class EpicsHandler extends BaseHttpHandler implements HttpHandler {
+    private final TaskManager manager;
+    private final Gson gson;
```
Переменную можно было бы перенести в базовый класс, она была бы доступна во всех классах наследниках и не надо было бы дублировать её. 
Один раз создать gson и использовать

avfyodorov commented on src/main/ru/kanban/http/handler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,36 @@
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+public class BaseHttpHandler {
```
Можно было бы провести рефакторинг, в базовом классе (BaseHttpHandler) оставить обработчик и методы-заглушки, а в классах наследниках переопределять только эти методы (processGet, processPost….), если они нужны. Такое решение лучше соответствует принципам OOP 
```java
public  class BaseHandler implements HttpHandler {
......................

public void handle(HttpExchange exchange) throws IOException { 
       String method = exchange.getRequestMethod();
           switch (method) {
               case "GET":
                   processGet(exchange);
                   break;
               case "POST":
                   processPost(exchange);
                   break;
               case "DELETE":
                   processDelete(exchange);
                   break;
               default:
                   writeToUser(exchange, "Данный метод не предусмотрен");
           }
           
    protected void processGet(....) {
        sendMethodNotAllowed(...);
   }
   protected void processPost(....) {
        sendMethodNotAllowed(...);
   }
.......................
И здесь же общие методы для классов-наследников. 

```
   Далее в каждом из наследников мы будем переопределять только нужные для этого наследника методы  processGet, processPost и т.д. и реализовывать в них нужную нам логику)

   На Ваше усмотрение.

avfyodorov commented on src/main/ru/kanban/http/handler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,36 @@
+    protected static final String TEXT_400 = "{\"error\":\"Некорректный формат ID в пути\"}";
+    protected static final String TEXT_404 = "{\"error\":\"Некорректный путь запроса\"}";
+    protected static final String TEXT_500 = "{\"error\":\"Ошибка сервера: %s\"}";
+
```
Для задач приблизительно так:
```java
public class TasksHandler extends BaseHandler {
......................
    public TasksHandler(........................
    
    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
  .........................
    @Override
    protected void processPost(HttpExchange exchange) throws IOException {
 ...............
    @Override
    protected void processDelete(HttpExchange exchange, String path) throws IOException {
.......................
```
.-----------------------------

А для отсортированного по времени списка достаточно будет переопределить всего один метод: 

```java
public class PrioritizedHandler extends BaseHandler {

    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
..................
```
.---------------------

Аналогично определяются классы-наследники и для эпиков, подзадач, истории.

avfyodorov reviewed (commented):
Добрый день, Максим! 

Очень хорошая работа, от меня всего пара небольших уточнений.

Commit:
Выполнил исправления по замечаниям ревьюера.

- Убрал handle из классов в BaseHttpHandler, теперь handle like классы переиспользуют его логику.
- Методы HTTP переопределены в классах.
- Убрал объявление и инициализацию Gson gson в BaseHttpHandler.

Commit:
Исправил баг в Adapter, из-за которого он принимал неправильно состояние taskManager.

- Вынес инициализацию gson на уровень выше, чтобы передавать одинаковое состояние taskManager и gson во все Handlers и Adapters.

avfyodorov reviewed (commented):
Добрый день, Максим!

Замечаний нет.
Работа принята.
